caffe2/opt/backend_cutting.cc

```diff
@@ -0,0 +1,416 @@
+    q.push(n);
+  }
+
+  while (!q.empty()) {
+    auto node = q.front();
+    q.pop();
+    auto& info = context->infos.at(node);
+
+    // Check if the node is supported, stop exploring further if not supported
+    auto* node_data = node->data().get();
+    if (isa<nom::repr::NeuralNetOperator>(node_data)) {
+      const auto* nn_op =
+          dyn_cast<const nom::repr::NeuralNetOperator>(node_data);
+      const auto* op_def = reinterpret_cast<const caffe2::OperatorDef*>(
+          nn_op->getAnnotation()->getSaved());
+      bool wanted = context->predicate(*op_def);
+      wanted = context->find_supported ? wanted : (!wanted);
+      if (!wanted) {
+        context->frontier.emplace(node);
+        continue;
+      }
+    }
+
+    // Adding to current group
+    info.group = context->group;
+    info.needs_transform = context->find_supported;
+    context->current_group.push_back(node);
+
+    // Continue exploring its fanouts
+    for (const auto& out_edge : node->getOutEdges()) {
+      auto child_node = out_edge->head();
+      auto& child_info = context->infos.at(child_node);
+      if (--child_info.in_degree == 0) {
+        q.push(child_node);
+      }
+    }
+  }
+}
+
+// Note: subgraph always starts with ops and ends with tensors, except for the
+// very first group, which can be all tensors
+struct TransformSubgraph {
+  explicit TransformSubgraph(
+      std::vector<NodeRef>&& f,
+      std::vector<NodeRef>&& n,
+      int id,
+      bool need)
+      : input_nodes(std::move(f)),
+        nodes(std::move(n)),
+        group_id(id),
+        needed(need) {}
+
+  TransformSubgraph(TransformSubgraph&& rhs) noexcept
+      : input_nodes(std::move(rhs.input_nodes)),
+        nodes(std::move(rhs.nodes)),
+        external_input_refs(std::move(rhs.external_input_refs)),
+        external_output_refs(std::move(rhs.external_output_refs)),
+        group_id(rhs.group_id),
+        needed(rhs.needed) {}
+
+  TransformSubgraph& operator=(TransformSubgraph&& rhs) noexcept {
+    input_nodes = std::move(rhs.input_nodes);
+    nodes = std::move(rhs.nodes);
+    external_input_refs = std::move(external_input_refs);
+    external_output_refs = std::move(external_output_refs);
+    group_id = rhs.group_id;
+    needed = rhs.needed;
+    return *this;
+  }
+
+  void Print() const {
+    LOG(INFO) << "Group :" << group_id;
+    LOG(INFO) << "  Input Nodes: ";
+    for (const auto i : input_nodes) {
+      LOG(INFO) << "    " << ShowNode(i);
+    }
+    LOG(INFO) << "  Nodes: ";
+    for (const auto i : nodes) {
+      LOG(INFO) << "    " << ShowNode(i);
+    }
+  }
+
+  std::vector<NodeRef> input_nodes;
+  std::vector<NodeRef> nodes;
+  std::unordered_map<std::string, NodeRef> external_input_refs;
+  std::unordered_map<std::string, NodeRef> external_output_refs;
+  int group_id{-1};
+  bool needed{true};
+};
+
+// Merge subgraph
+// Precondition: subgraphs appear in topological order in the list
+void MergeSubgraph(
```

Comment:
no-op?

Reply:
Yes, I figured that it's not needed. Will remove this part.